### PR TITLE
Check packet events have not been handled before processing

### PR DESCRIPTION
**Pull request by ancazamfir**
Closes: #983

## Description
- check that a send packet has not been handled before processing
   - checks that the packet has not been received on the destination chain and 
   -  the packet commitment is still present on source chain (i.e. the packet was not timed out)
- check  that a packet received and acknowledged on destination chain has not been already handled
   - checks that an acknowledgment packet has not been received by the source chain 

~~(ignore the changes in the client worker...they are coming from PR #974 and will not be committed)~~

Testing:
- manually tested running multiple relayers with different accounts. Send a number of packets with high and small timeout offsets
- slower instance retries only once before `already handled` (found in logs) event is detected
- performed the tests for recv, ack and timeout packets

______

For contributor use:

- [x] Updated the __Unreleased__ section of [CHANGELOG.md](https://github.com/informalsystems/ibc-rs/blob/master/CHANGELOG.md) with the issue.
- [ ] If applicable: Unit tests written, added test to CI.
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Updated relevant documentation (`docs/`) and code comments.
- [ ] Re-reviewed `Files changed` in the Github PR explorer.

### CODE REVIEW

File: modules/src/ics04_channel/packet.rs [hunk 1]
```diff
@@ -112,6 +113,14 @@ pub struct Packet {
     pub timeout_timestamp: Timestamp,
 }
 
+impl Packet {
+    pub fn timed_out(&self, dst_chain_height: Height) -> bool {
+        self.timeout_height != Height::zero() && self.timeout_height < dst_chain_height
+            || self.timeout_timestamp != Timestamp::none()
+                && Timestamp::now().check_expiry(&self.timeout_timestamp) == Expired
+    }
```
Comment (adizere) [inline modules/src/ics04_channel/packet.rs:121]: The order of precedence here is not obvious. Could we improve by making precedence explicit? I am guessing it should be like this:

```suggestion
    pub fn timed_out(&self, dst_chain_height: Height) -> bool {
        (self.timeout_height != Height::zero() && self.timeout_height < dst_chain_height)
            || (self.timeout_timestamp != Timestamp::none()
                && Timestamp::now().check_expiry(&self.timeout_timestamp)) == Expired
    }
```

File: modules/src/ics04_channel/packet.rs [hunk 2]
```diff
@@ -14,6 +14,7 @@ use super::handler::{
     acknowledgement::AckPacketResult, recv_packet::RecvPacketResult, send_packet::SendPacketResult,
     timeout::TimeoutPacketResult, write_acknowledgement::WriteAckPacketResult,
 };
+use crate::timestamp::Expiry::Expired;
```
Comment (adizere) [inline modules/src/ics04_channel/packet.rs:17]: SuperNit: `crate` imports are higher.

File: relayer/src/link.rs [hunk 3]
```diff
@@ -714,6 +722,57 @@ impl RelayPath {
...
+        Ok(bytes.is_empty())
+    }
+
+    /// Checks if a send packet event has already been handled (e.g. by another relayer).
+    fn send_packet_event_handled(&self, sp: &SendPacket) -> Result<bool, LinkError> {
+        Ok(self.send_packet_received_on_dst(&sp.packet)?
+            || self.send_packet_commitment_cleared_on_src(&sp.packet)?)
+    }
+
+    /// Checks if an acknowledge packet has been received on source chain.
```
Comment (adizere) [inline relayer/src/link.rs:758]: Nit: This method can be confusing. Both the comment and method name talk about the source chain, but the query is against the destination chain.

**Review by adizere — COMMENTED** [review]
Left some minor comments. I am still testing this..

File: modules/src/ics04_channel/packet.rs (hunk 1 above)
Comment (romac) [reply modules/src/ics04_channel/packet.rs]: b08c348b

File: modules/src/ics04_channel/packet.rs (hunk 2 above)
Comment (romac) [reply modules/src/ics04_channel/packet.rs]: 62454bdf

File: relayer/src/link.rs (hunk 3 above)
Comment (ancazamfir) [reply relayer/src/link.rs:758]: its the packet's source chain...which is the destination chain for the relay path that sends the ack. Definitely the comment should be clearer than it is.

Comment (romac) [reply relayer/src/link.rs]: f915d886c2cb2ffcf006c8ac451b97055e09cb45

**Review by adizere — APPROVED** [review]

**Pull request merged**